addons/point_of_sale/static/src/app/models/product_product.js

```diff
@@ -0,0 +1,224 @@
+ * with the key "pos_available_models". And to be instanciated with the
+ * method createRelatedModels from related_models.js
+ *
+ * Models to load: product.product, uom.uom
+ */
+
+export class ProductProduct extends Base {
+    static pythonModel = "product.product";
+
+    // Getter
+    isAllowOnlyOneLot() {
+        const productUnit = this.uom_id;
+        return this.tracking === "lot" || !productUnit || !productUnit.is_pos_groupable;
+    }
+
+    isTracked() {
+        const stock_picking_type = this.models["stock.picking.type"].readAll()[0];
+
+        return (
+            ["serial", "lot"].includes(this.tracking) &&
+            (stock_picking_type.use_create_lots || stock_picking_type.use_existing_lots)
+        );
+    }
+
+    async _onScaleNotAvailable() {}
+
+    isConfigurable() {
+        const attrByIds = this.models["product.attribute"].getAllBy("id");
+        return this.attribute_line_ids.some((line) => line.attribute_id?.id in attrByIds);
+    }
+
+    get isScaleAvailable() {
+        return true;
+    }
+
+    get parentCategoryIds() {
+        const categories = [];
+        let category = this.categ_id;
+
+        while (category) {
+            categories.push(category.id);
+            category = category.parent_id;
+        }
+
+        return categories;
+    }
+
+    get parentPosCategIds() {
+        const current = [];
+        const categories = this.pos_categ_ids;
+
+        const getParent = (categ) => {
+            if (categ.parent_id) {
+                current.push(categ.parent_id.id);
+                getParent(categ.parent_id);
+            }
+        };
+
+        for (const category of categories) {
+            current.push(category.id);
+            getParent(category);
+        }
+
+        return current;
+    }
+
+    get childPosCategIds() {
+        const current = [];
+        const categories = this.pos_categ_ids;
+
+        const getChild = (categ) => {
+            if (categ.child_id) {
+                for (const child of categ.child_id) {
+                    current.push(child.id);
+                    getChild(child);
+                }
+            }
+        };
+
+        for (const category of categories) {
+            current.push(category.id);
+            getChild(category);
+        }
+
+        return current;
+    }
+
+    get parent_pos_categ_ids() {
+        const categories = [];
+        let category = this.pos_categ_id;
+
+        while (category) {
+            categories.push(category.id);
+            category = category.parent_id;
+        }
+
+        return categories;
+    }
+
+    isPricelistItemUsable(item, date) {
+        return (
+            (!item.categ_id || this.parentCategoryIds.includes(item.categ_id.id)) &&
+            (!item.date_start || deserializeDate(item.date_start) <= date) &&
+            (!item.date_end || deserializeDate(item.date_end) >= date)
+        );
+    }
+
+    get_pricelist_rule(pricelist) {
```

Comment:
don't forget to take into account the optimization around `applicablePricelistItems` which was introduced because pos can be loaded with pricelists containing pricelist items (rules) corresponding to each loaded product.